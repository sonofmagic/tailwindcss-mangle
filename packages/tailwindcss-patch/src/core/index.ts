export * from './tw-patcher'
export * from './exposeContext'
export * from './inspector'
export * from './runtime-patcher'
export * from './cache'
export * from './config'
