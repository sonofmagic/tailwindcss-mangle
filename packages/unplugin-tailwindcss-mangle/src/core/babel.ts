import babel from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'
import MagicString from 'magic-string'
import { splitCode } from '@tailwindcss-mangle/shared'
interface Options {
  replaceMap: Map<string, string>
  magicString: MagicString
}

export const plugin = declare((api, options: Options) => {
  api.assertVersion(7)
  const { magicString, replaceMap } = options
  return {
    visitor: {
      StringLiteral: {
        enter(p) {
          const node = p.node
          let value = node.value
          const arr = splitCode(value)

          for (const str of arr) {
            if (replaceMap.has(str)) {
              const v = replaceMap.get(str)
              if (v) {
                value = value.replaceAll(str, v)
              }
            }
          }
          if (typeof node.start === 'number' && typeof node.end === 'number' && value) {
            magicString.update(node.start + 1, node.end - 1, value)
          }
        }
      }
    }
  }
})

export function processJs(options: { code: string; replaceMap: Map<string, string> }) {
  const { code, replaceMap } = options
  const magicString = new MagicString(code)
  babel.transformSync(code, {
    presets: [
      // ['@babel/preset-react', {}],
      [
        '@babel/preset-typescript',
        {
          allExtensions: true,
          isTSX: true
        }
      ]
    ],
    plugins: [
      [
        plugin,
        {
          magicString,
          replaceMap
        }
      ]
    ]
  })
  return magicString.toString()
}
