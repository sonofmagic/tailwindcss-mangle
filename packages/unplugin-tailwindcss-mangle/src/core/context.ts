import fs from 'node:fs'
import { resolve } from 'node:path'
import { ClassGenerator } from '@tailwindcss-mangle/shared'
import { getConfig, getDefaultMangleUserConfig } from '@tailwindcss-mangle/config'
import type { MangleUserConfig } from '@tailwindcss-mangle/config'
import { sort } from 'fast-sort'
import defu from 'defu'
import { createGlobMatcher, defaultMangleClassFilter } from '@/utils'

export class Context {
  options: MangleUserConfig
  includeMatcher: (file: string) => boolean
  excludeMatcher: (file: string) => boolean
  classSet: Set<string>
  replaceMap: Map<string, string>
  classGenerator: ClassGenerator
  constructor(opts: MangleUserConfig) {
    this.options = defu(opts, getDefaultMangleUserConfig())
    this.classSet = new Set()
    this.replaceMap = new Map()
    this.includeMatcher = createGlobMatcher(this.options.include, true)
    this.excludeMatcher = createGlobMatcher(this.options.exclude, false)
    this.classGenerator = new ClassGenerator(this.options.classGenerator)
  }

  mergeOptions(opts?: MangleUserConfig) {
    // 插件选项优先
    this.options = defu(this.options, opts)
    this.includeMatcher = createGlobMatcher(this.options.include, true)
    this.excludeMatcher = createGlobMatcher(this.options.exclude, false)
    this.classGenerator = new ClassGenerator(this.options.classGenerator)
  }

  isInclude(file: string) {
    return this.includeMatcher(file) && !this.excludeMatcher(file)
  }

  currentMangleClassFilter(className: string) {
    return (this.options.mangleClassFilter ?? defaultMangleClassFilter)(className)
  }

  getClassSet() {
    return this.classSet
  }

  getReplaceMap() {
    return this.replaceMap
  }

  addToUsedBy(key: string, file: string) {
    const hit = this.classGenerator.newClassMap[key]
    if (hit) {
      hit.usedBy.add(file)
    }
  }

  async initConfig() {
    const { config } = await getConfig()
    const mangleConfig = config?.mangle
    this.mergeOptions(mangleConfig)
    const jsonPath = this.options.classListPath ?? resolve(process.cwd(), config?.patch?.output?.filename as string)

    if (jsonPath && fs.existsSync(jsonPath)) {
      const rawClassList = fs.readFileSync(jsonPath, 'utf8')
      const list = JSON.parse(rawClassList) as string[]
      // why?
      // case bg-red-500 and bg-red-500/50
      // transform bg-red-500/50 first
      const classList = sort(list).desc((c) => c.length)
      for (const className of classList) {
        if (this.currentMangleClassFilter(className)) {
          this.classSet.add(className)
        }
      }
    }
    for (const cls of this.classSet) {
      this.classGenerator.generateClassName(cls)
    }

    for (const x of Object.entries(this.classGenerator.newClassMap)) {
      this.replaceMap.set(x[0], x[1].name)
    }
    return config
  }
}
