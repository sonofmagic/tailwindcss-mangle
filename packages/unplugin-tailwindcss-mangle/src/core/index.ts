export * from './plugin'
export * from './context'
