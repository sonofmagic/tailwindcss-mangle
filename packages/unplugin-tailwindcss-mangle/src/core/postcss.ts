import postcss, { type PluginCreator } from 'postcss'
import parser from 'postcss-selector-parser'

const transformSelectorPostcssPlugin: PluginCreator<{
  replaceMap: Map<string, string>
}> = function (options) {
  const { replaceMap } = options ?? {}
  return {
    postcssPlugin: 'postcssTransformSelectorPlugin',
    async Rule(rule) {
      await parser((selectors) => {
        selectors.walkClasses((s) => {
          if (s.value && replaceMap && replaceMap.has(s.value)) {
            const v = replaceMap.get(s.value)
            if (v) {
              s.value = v
            }
          }
        })
      }).transform(rule, {
        lossless: false,
        updateSelector: true
      })
    }
  }
}
transformSelectorPostcssPlugin.postcss = true

export function processCss(options: { replaceMap: Map<string, string>; css: string; file: string }) {
  const { css, file, replaceMap } = options
  return postcss([
    transformSelectorPostcssPlugin({
      replaceMap
    })
  ]).process(css, {
    from: file,
    to: file
  })
}
