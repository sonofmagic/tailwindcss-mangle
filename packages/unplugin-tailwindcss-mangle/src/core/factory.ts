import type { TransformerOptions } from '@tailwindcss-mangle/config'
import type { UnpluginFactory } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import { Context, cssHandler, htmlHandler, jsHandler } from '@tailwindcss-mangle/core'
import { isCSSRequest } from 'is-css-request'
import path from 'pathe'
import { getGroupedEntries } from '@/utils'
import { pluginName } from '../constants'

const WEBPACK_LOADER = path.resolve(__dirname, __DEV__ ? '../../dist/loader.cjs' : './loader.cjs')

const factory: UnpluginFactory<TransformerOptions | undefined> = (options) => {
  const ctx = new Context()
  let filter = (_id: string) => true
  return [
    {
      name: `${pluginName}:pre`,
      enforce: 'pre',
      async buildStart() {
        await ctx.initConfig({
          transformerOptions: options,
        })
        filter = createFilter(ctx.options.sources?.include, ctx.options.sources?.exclude)
      },
    },
    {
      name: `${pluginName}`,
      transformInclude(id) {
        return filter(id)
      },
      async transform(code, id) {
        const opts = {
          ctx,
          id,
        }
        if (/\.[cm]?[jt]sx?(?:$|\?)/.test(id)) {
          return jsHandler(code, opts)
        }
        else if (/\.(?:vue|svelte)(?:$|\?)/.test(id)) {
          if (isCSSRequest(id)) {
            return await cssHandler(code, opts)
          }
          else {
            return jsHandler(code, opts)
          }
        }
        else if (isCSSRequest(id)) {
          return await cssHandler(code, opts)
        }
        else if (/\.html?/.test(id)) {
          return htmlHandler(code, opts)
        }
      },
      webpack(compiler) {
        const { NormalModule } = compiler.webpack
        const isExisted = true
        compiler.hooks.compilation.tap(pluginName, (compilation) => {
          NormalModule.getCompilationHooks(compilation).loader.tap(pluginName, (_loaderContext, module) => {
            if (isExisted) {
              const idx = module.loaders.findIndex(x => x.loader.includes('postcss-loader'))

              if (idx > -1) {
                module.loaders.splice(idx, 0, {
                  loader: WEBPACK_LOADER,
                  ident: null,
                  options: {
                    ctx,
                  },
                  type: null,
                })
              }
            }
          })
        })
      },
    },
    {
      name: `${pluginName}:post`,
      enforce: 'post',
      vite: {
        transformIndexHtml(html) {
          const { code } = htmlHandler(html, { ctx })
          return code
        },
        // generateBundle: {
        //   async handler(options, bundle) {
        //     const groupedEntries = getGroupedEntries(Object.entries(bundle))

        //     if (Array.isArray(groupedEntries.css) && groupedEntries.css.length > 0) {
        //       for (let i = 0; i < groupedEntries.css.length; i++) {
        //         const [id, cssSource] = groupedEntries.css[i] as [string, OutputAsset]

        //         const { code } = await cssHandler(cssSource.source.toString(), {
        //           id,
        //           ctx,
        //         })
        //         cssSource.source = code
        //       }
        //     }
        //   },
        // },
      },
      webpack(compiler) {
        const { Compilation, sources } = compiler.webpack
        const { ConcatSource } = sources

        compiler.hooks.compilation.tap(pluginName, (compilation) => {
          compilation.hooks.processAssets.tapPromise(
            {
              name: pluginName,
              stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE,
            },
            async (assets) => {
              const groupedEntries = getGroupedEntries(Object.entries(assets))

              // if (groupedEntries.js.length > 0) {
              //   for (let i = 0; i < groupedEntries.js.length; i++) {
              //     const [file, chunk] = groupedEntries.js[i]

              //     const code = jsHandler(chunk.source().toString(), {
              //       ctx,
              //     }).code
              //     if (code) {
              //       const source = new ConcatSource(code)
              //       compilation.updateAsset(file, source)
              //     }
              //   }
              // }

              if (groupedEntries.css.length > 0) {
                for (let i = 0; i < groupedEntries.css.length; i++) {
                  const [id, cssSource] = groupedEntries.css[i]

                  const { code } = await cssHandler(cssSource.source().toString(), {
                    id,
                    ctx,
                  })

                  const source = new ConcatSource(code)

                  compilation.updateAsset(id, source)
                }
              }

              // if (groupedEntries.html.length > 0) {
              //   for (let i = 0; i < groupedEntries.html.length; i++) {
              //     const [file, asset] = groupedEntries.html[i]

              //     const { code } = htmlHandler(asset.source().toString(), {
              //       ctx,
              //     })
              //     const source = new ConcatSource(code)
              //     compilation.updateAsset(file, source)
              //   }
              // }
            },
          )
        })
      },
      writeBundle() {
        ctx.dump()
      },
    },
  ]
}

export default factory
