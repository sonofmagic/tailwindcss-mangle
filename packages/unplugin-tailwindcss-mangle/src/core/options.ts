import fs from 'node:fs'
import { resolve } from 'node:path'
import { ClassGenerator } from '@tailwindcss-mangle/shared'
import { getConfig, getDefaultUserConfig } from '@tailwindcss-mangle/config'
import type { UserConfig } from '@tailwindcss-mangle/config'
import defu from 'defu'
import { sort } from 'fast-sort'
import { htmlHandler, cssHandler, jsHandler, preProcessJs } from '@tailwindcss-mangle/core'
import type { Options, ClassMapOutputOptions } from '@/types'
import { createGlobMatcher, defaultMangleClassFilter } from '@/utils'

export function getOptions(opts: Options | undefined = {}) {
  const {
    include,
    exclude,
    disabled,
    mangleClassFilter,
    classMapOutput,
    classGenerator: classGeneratorOptions,
    classListPath: _classListPath
  } = defu<Options, Options[]>(opts, {
    include: ['**/*.{js,jsx,ts,tsx,svelte,vue}'],
    exclude: ['**/*.{css,scss,less,sass,postcss,html,htm}'],
    disabled: process.env.NODE_ENV === 'development'
  })
  const includeMatcher = createGlobMatcher(include, true)
  const excludeMatcher = createGlobMatcher(exclude, false)
  const currentMangleClassFilter = mangleClassFilter ?? defaultMangleClassFilter

  function isInclude(file: string) {
    return includeMatcher(file) && !excludeMatcher(file)
  }

  const classSet: Set<string> = new Set()
  const replaceMap: Map<string, string> = new Map()
  let userConfig: UserConfig = getDefaultUserConfig()

  const classMapOutputOptions: ClassMapOutputOptions = {
    filename: 'classMap.json'
  }

  if (typeof classMapOutput === 'object') {
    Object.assign(classMapOutputOptions, classMapOutput)
  }

  // let cached: boolean
  const classGenerator = new ClassGenerator(classGeneratorOptions)

  function getCachedClassSet() {
    return classSet
  }

  function getReplaceMap() {
    return replaceMap
  }

  async function initConfig() {
    const { config } = await getConfig()
    userConfig = config as UserConfig
    let classListPath: string = ''
    if (userConfig) {
      classListPath = resolve(process.cwd(), userConfig.patch?.output?.filename as string)
    }
    if (_classListPath) {
      classListPath = _classListPath
    }

    if (classListPath && fs.existsSync(classListPath)) {
      const rawClassList = fs.readFileSync(classListPath, 'utf8')
      const list = JSON.parse(rawClassList) as string[]
      // why?
      // case bg-red-500 and bg-red-500/50
      // transform bg-red-500/50 first
      const classList = sort(list).desc((c) => c.length)
      for (const className of classList) {
        if (currentMangleClassFilter(className)) {
          classSet.add(className)
        }
      }
    }
    for (const cls of classSet) {
      classGenerator.generateClassName(cls)
    }

    for (const x of Object.entries(classGenerator.newClassMap)) {
      replaceMap.set(x[0], x[1].name)
    }
    return config
  }

  function addToUsedBy(key: string, file: string) {
    const hit = classGenerator.newClassMap[key]
    if (hit) {
      hit.usedBy.add(file)
    }
  }

  return {
    getCachedClassSet,
    classGenerator,
    includeMatcher,
    excludeMatcher,
    isInclude,
    classMapOutputOptions,
    initConfig,
    getReplaceMap,
    addToUsedBy,
    disabled,
    htmlHandler,
    cssHandler,
    jsHandler,
    preProcessJs
  }
}
