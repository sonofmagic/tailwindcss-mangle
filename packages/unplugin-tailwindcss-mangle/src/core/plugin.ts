import { dirname } from 'node:path'
import fs from 'node:fs/promises'
import { createUnplugin } from 'unplugin'
import type { OutputAsset } from 'rollup'
import { Context, cssHandler, htmlHandler, jsHandler, preProcessJs, preProcessRawCode } from '@tailwindcss-mangle/core'
import type { ClassMapOutputOptions, MangleUserConfig } from '@tailwindcss-mangle/config'
import MagicString from 'magic-string'
import { pluginName } from '@/constants'
import { ensureDir, getGroupedEntries } from '@/utils'

export const unplugin = createUnplugin((options?: MangleUserConfig) => {
  const ctx = new Context()

  return {
    name: pluginName,
    enforce: 'pre',

    async buildStart() {
      await ctx.initConfig({
        mangleOptions: options,
      })
    },
    transformInclude(id) {
      return ctx.isInclude(id)
    },
    transform(code, id) {
      const s = new MagicString(code)
      const replaceMap = ctx.getReplaceMap()
      // 直接忽略 css  文件，因为此时 tailwindcss 还没有展开
      return /\.[jt]sx?$/.test(id)
        ? preProcessJs({
          code: s,
          replaceMap,
          ctx,
          id,
        })
        : preProcessRawCode({
          code,
          ctx,
          replaceMap,
          id,
        })
    },
    vite: {
      generateBundle: {
        async handler(options, bundle) {
          const replaceMap = ctx.getReplaceMap()
          const groupedEntries = getGroupedEntries(Object.entries(bundle))

          if (Array.isArray(groupedEntries.css) && groupedEntries.css.length > 0) {
            for (let i = 0; i < groupedEntries.css.length; i++) {
              const [file, cssSource] = groupedEntries.css[i] as [string, OutputAsset]

              const { css } = await cssHandler(cssSource.source.toString(), {
                file,
                replaceMap,
                ctx,
              })
              cssSource.source = css
            }
          }
        },
      },
    },
    webpack(compiler) {
      const { Compilation, sources } = compiler.webpack
      const { ConcatSource } = sources

      compiler.hooks.compilation.tap(pluginName, (compilation) => {
        compilation.hooks.processAssets.tapPromise(
          {
            name: pluginName,
            stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE,
          },
          async (assets) => {
            const replaceMap = ctx.getReplaceMap()
            const groupedEntries = getGroupedEntries(Object.entries(assets))

            if (groupedEntries.js.length > 0) {
              for (let i = 0; i < groupedEntries.js.length; i++) {
                const [file, chunk] = groupedEntries.js[i]

                const code = jsHandler(chunk.source().toString(), {
                  replaceMap,
                  ctx,
                }).code
                if (code) {
                  const source = new ConcatSource(code)
                  compilation.updateAsset(file, source)
                }
              }
            }

            if (groupedEntries.css.length > 0) {
              for (let i = 0; i < groupedEntries.css.length; i++) {
                const [file, cssSource] = groupedEntries.css[i]

                const { css } = await cssHandler(cssSource.source().toString(), {
                  replaceMap,
                  file,
                  ctx,
                })

                const source = new ConcatSource(css)

                compilation.updateAsset(file, source)
              }
            }

            if (groupedEntries.html.length > 0) {
              for (let i = 0; i < groupedEntries.html.length; i++) {
                const [file, asset] = groupedEntries.html[i]

                const html = htmlHandler(asset.source().toString(), {
                  ctx,
                  replaceMap,
                })
                const source = new ConcatSource(html)
                compilation.updateAsset(file, source)
              }
            }
          },
        )
      })
    },
    async writeBundle() {
      if (ctx.options.classMapOutput?.enable) {
        const opts = ctx.options.classMapOutput as Required<ClassMapOutputOptions>
        const entries = Object.entries(ctx.classGenerator.newClassMap)
        if (entries.length > 0 && opts) {
          await ensureDir(dirname(opts.filename))
          const output = JSON.stringify(
            entries.map((x) => {
              return {
                origin: x[0],
                replacement: x[1].name,
                usedBy: [...x[1].usedBy],
              }
            }),
            null,
            opts.loose ? 2 : 0,
          )
          await fs.writeFile(opts.filename, output, 'utf8')
          console.log(`✨ ${opts.filename} generated!`)
        }
      }
    },
  }
})
