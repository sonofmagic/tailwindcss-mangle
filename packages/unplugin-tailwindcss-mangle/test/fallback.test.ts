import fs from 'node:fs/promises'
import path from 'node:path'
// import MagicString from 'magic-string'
import normalizeNewline from 'normalize-newline'
import { Context } from '@/core/context'
const fallbackDir = path.resolve(__dirname, 'fixtures/fallback')
describe('fallback', () => {
  it('ac', async () => {
    const ctx = new Context()
    await ctx.initConfig(fallbackDir)
    const code = await fs.readFile(path.resolve(fallbackDir, 'index.vue'), 'utf8')
    const { arr } = ctx.search(normalizeNewline(code))
    // const s = new MagicString(code)
    // const replaceMap = ctx.getReplaceMap()
    expect(arr).toMatchSnapshot()
    // for (const [cls, ends] of arr) {
    //   for (const end of ends) {
    //     const value = replaceMap.get(cls)
    //     if (value) {
    //       const res = s.slice(end - cls.length + 1, end + 1)
    //       expect(res).toBe(cls)
    //     }
    //   }
    // }
    // for (const [start, strs] of arr) {
    //   for (const str of strs) {
    //     const value = replaceMap.get(str)
    //     if (value) {
    //       const res = s.slice(start - str.length + 1, start + 1)
    //       expect(res).toBe(str)
    //       console.log(str, value)

    //       // s.update(start, start + str.length, value)
    //     }
    //   }
    // }
  })
})
