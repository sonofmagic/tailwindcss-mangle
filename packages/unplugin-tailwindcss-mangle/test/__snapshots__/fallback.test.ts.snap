// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fallback > ac 1`] = `
[
  [
    "dark:drop-shadow-[0_0_0.3rem_#ffffff70]",
    [
      [
        1855,
        1894,
      ],
    ],
  ],
  [
    "before:dark:bg-gradient-to-br",
    [
      [
        1668,
        1697,
      ],
    ],
  ],
  [
    "hover:dark:border-neutral-700",
    [
      [
        2325,
        2354,
      ],
      [
        3035,
        3064,
      ],
      [
        3774,
        3803,
      ],
      [
        4492,
        4521,
      ],
    ],
  ],
  [
    "before:dark:from-transparent",
    [
      [
        1698,
        1726,
      ],
    ],
  ],
  [
    "hover:dark:bg-neutral-800/30",
    [
      [
        2355,
        2383,
      ],
      [
        3065,
        3093,
      ],
      [
        3804,
        3832,
      ],
      [
        4522,
        4550,
      ],
    ],
  ],
  [
    "motion-reduce:transform-none",
    [
      [
        2585,
        2613,
      ],
      [
        3296,
        3324,
      ],
      [
        4039,
        4067,
      ],
      [
        4754,
        4782,
      ],
    ],
  ],
  [
    "after:dark:via-[#0141ff]/40",
    [
      [
        1778,
        1805,
      ],
    ],
  ],
  [
    "before:dark:to-blue-700/10",
    [
      [
        1727,
        1753,
      ],
    ],
  ],
  [
    "group-hover:translate-x-1",
    [
      [
        2559,
        2584,
      ],
      [
        3270,
        3295,
      ],
      [
        4013,
        4038,
      ],
      [
        4728,
        4753,
      ],
    ],
  ],
  [
    "dark:border-neutral-800",
    [
      [
        338,
        361,
      ],
    ],
  ],
  [
    "before:-translate-x-1/2",
    [
      [
        1345,
        1368,
      ],
    ],
  ],
  [
    "after:dark:from-sky-900",
    [
      [
        1754,
        1777,
      ],
    ],
  ],
  [
    "lg:dark:bg-zinc-800/30",
    [
      [
        467,
        489,
      ],
    ],
  ],
  [
    "lg:pointer-events-auto",
    [
      [
        879,
        901,
      ],
    ],
  ],
  [
    "before:to-transparent",
    [
      [
        1433,
        1454,
      ],
    ],
  ],
  [
    "after:translate-x-1/3",
    [
      [
        1550,
        1571,
      ],
    ],
  ],
  [
    "hover:border-gray-300",
    [
      [
        2285,
        2306,
      ],
      [
        2995,
        3016,
      ],
      [
        3734,
        3755,
      ],
      [
        4452,
        4473,
      ],
    ],
  ],
  [
    "transition-transform",
    [
      [
        2538,
        2558,
      ],
      [
        3249,
        3269,
      ],
      [
        3992,
        4012,
      ],
      [
        4707,
        4727,
      ],
    ],
  ],
  [
    "dark:bg-zinc-800/30",
    [
      [
        362,
        381,
      ],
      [
        470,
        489,
      ],
    ],
  ],
  [
    "pointer-events-none",
    [
      [
        825,
        844,
      ],
    ],
  ],
  [
    "before:rounded-full",
    [
      [
        1369,
        1388,
      ],
    ],
  ],
  [
    "before:content-['']",
    [
      [
        1471,
        1490,
      ],
    ],
  ],
  [
    "before:lg:h-[360px]",
    [
      [
        1806,
        1825,
      ],
    ],
  ],
  [
    "place-items-center",
    [
      [
        850,
        868,
      ],
      [
        1276,
        1294,
      ],
    ],
  ],
  [
    "after:from-sky-200",
    [
      [
        1596,
        1614,
      ],
    ],
  ],
  [
    "after:via-blue-200",
    [
      [
        1615,
        1633,
      ],
    ],
  ],
  [
    "after:content-['']",
    [
      [
        1649,
        1667,
      ],
    ],
  ],
  [
    "border-transparent",
    [
      [
        2238,
        2256,
      ],
      [
        2948,
        2966,
      ],
      [
        3687,
        3705,
      ],
      [
        4405,
        4423,
      ],
    ],
  ],
  [
    "backdrop-blur-2xl",
    [
      [
        320,
        337,
      ],
    ],
  ],
  [
    "dark:from-inherit",
    [
      [
        382,
        399,
      ],
    ],
  ],
  [
    "before:from-white",
    [
      [
        1415,
        1432,
      ],
    ],
  ],
  [
    "transition-colors",
    [
      [
        2267,
        2284,
      ],
      [
        2977,
        2994,
      ],
      [
        3716,
        3733,
      ],
      [
        4434,
        4451,
      ],
    ],
  ],
  [
    "hover:bg-gray-100",
    [
      [
        2307,
        2324,
      ],
      [
        3017,
        3034,
      ],
      [
        3756,
        3773,
      ],
      [
        4474,
        4491,
      ],
    ],
  ],
  [
    "bg-gradient-to-b",
    [
      [
        279,
        295,
      ],
      [
        1680,
        1696,
      ],
    ],
  ],
  [
    "bg-gradient-to-t",
    [
      [
        695,
        711,
      ],
    ],
  ],
  [
    "before:h-[300px]",
    [
      [
        1311,
        1327,
      ],
    ],
  ],
  [
    "before:w-[480px]",
    [
      [
        1328,
        1344,
      ],
    ],
  ],
  [
    "justify-between",
    [
      [
        66,
        81,
      ],
      [
        140,
        155,
      ],
    ],
  ],
  [
    "border-gray-300",
    [
      [
        263,
        278,
      ],
      [
        2291,
        2306,
      ],
      [
        3001,
        3016,
      ],
      [
        3740,
        3755,
      ],
      [
        4458,
        4473,
      ],
    ],
  ],
  [
    "dark:from-black",
    [
      [
        733,
        748,
      ],
    ],
  ],
  [
    "before:absolute",
    [
      [
        1295,
        1310,
      ],
    ],
  ],
  [
    "before:blur-2xl",
    [
      [
        1455,
        1470,
      ],
    ],
  ],
  [
    "after:h-[180px]",
    [
      [
        1518,
        1533,
      ],
    ],
  ],
  [
    "after:w-[240px]",
    [
      [
        1534,
        1549,
      ],
    ],
  ],
  [
    "justify-center",
    [
      [
        239,
        253,
      ],
      [
        680,
        694,
      ],
    ],
  ],
  [
    "lg:bg-gray-200",
    [
      [
        445,
        459,
      ],
    ],
  ],
  [
    "dark:via-black",
    [
      [
        749,
        763,
      ],
    ],
  ],
  [
    "after:absolute",
    [
      [
        1491,
        1505,
      ],
    ],
  ],
  [
    "after:blur-2xl",
    [
      [
        1634,
        1648,
      ],
    ],
  ],
  [
    "lg:grid-cols-4",
    [
      [
        2042,
        2056,
      ],
    ],
  ],
  [
    "from-zinc-200",
    [
      [
        296,
        309,
      ],
    ],
  ],
  [
    "lg:rounded-xl",
    [
      [
        421,
        434,
      ],
    ],
  ],
  [
    "font-semibold",
    [
      [
        2469,
        2482,
      ],
      [
        3179,
        3192,
      ],
      [
        3918,
        3931,
      ],
      [
        4636,
        4649,
      ],
    ],
  ],
  [
    "min-h-screen",
    [
      [
        31,
        43,
      ],
    ],
  ],
  [
    "items-center",
    [
      [
        53,
        65,
      ],
      [
        127,
        139,
      ],
      [
        856,
        868,
      ],
      [
        1282,
        1294,
      ],
    ],
  ],
  [
    "lg:text-left",
    [
      [
        2057,
        2069,
      ],
    ],
  ],
  [
    "inline-block",
    [
      [
        2525,
        2537,
      ],
      [
        3236,
        3248,
      ],
      [
        3979,
        3991,
      ],
      [
        4694,
        4706,
      ],
    ],
  ],
  [
    "max-w-[30ch]",
    [
      [
        2664,
        2676,
      ],
      [
        3375,
        3387,
      ],
      [
        4118,
        4130,
      ],
      [
        4833,
        4845,
      ],
    ],
  ],
  [
    "dark:invert",
    [
      [
        1177,
        1188,
      ],
      [
        1895,
        1906,
      ],
    ],
  ],
  [
    "after:-z-20",
    [
      [
        1506,
        1517,
      ],
    ],
  ],
  [
    "text-center",
    [
      [
        2022,
        2033,
      ],
    ],
  ],
  [
    "from-white",
    [
      [
        712,
        722,
      ],
      [
        1422,
        1432,
      ],
    ],
  ],
  [
    "lg:bg-none",
    [
      [
        794,
        804,
      ],
    ],
  ],
  [
    "rounded-lg",
    [
      [
        2220,
        2230,
      ],
      [
        2930,
        2940,
      ],
      [
        3669,
        3679,
      ],
      [
        4387,
        4397,
      ],
    ],
  ],
  [
    "opacity-50",
    [
      [
        2685,
        2695,
      ],
      [
        3396,
        3406,
      ],
      [
        4139,
        4149,
      ],
      [
        4854,
        4864,
      ],
    ],
  ],
  [
    "max-w-5xl",
    [
      [
        117,
        126,
      ],
    ],
  ],
  [
    "font-mono",
    [
      [
        156,
        165,
      ],
      [
        550,
        559,
      ],
    ],
  ],
  [
    "lg:static",
    [
      [
        400,
        409,
      ],
      [
        764,
        773,
      ],
    ],
  ],
  [
    "lg:w-auto",
    [
      [
        410,
        419,
      ],
      [
        784,
        793,
      ],
    ],
  ],
  [
    "lg:border",
    [
      [
        435,
        444,
      ],
    ],
  ],
  [
    "font-bold",
    [
      [
        560,
        569,
      ],
    ],
  ],
  [
    "items-end",
    [
      [
        670,
        679,
      ],
    ],
  ],
  [
    "via-white",
    [
      [
        723,
        732,
      ],
    ],
  ],
  [
    "lg:h-auto",
    [
      [
        774,
        783,
      ],
    ],
  ],
  [
    "flex-col",
    [
      [
        44,
        52,
      ],
    ],
  ],
  [
    "border-b",
    [
      [
        254,
        262,
      ],
    ],
  ],
  [
    "bottom-0",
    [
      [
        637,
        645,
      ],
    ],
  ],
  [
    "text-2xl",
    [
      [
        2460,
        2468,
      ],
      [
        3170,
        3178,
      ],
      [
        3909,
        3917,
      ],
      [
        4627,
        4635,
      ],
    ],
  ],
  [
    "text-sm",
    [
      [
        166,
        173,
      ],
      [
        2677,
        2684,
      ],
      [
        3388,
        3395,
      ],
      [
        4131,
        4138,
      ],
      [
        4846,
        4853,
      ],
    ],
  ],
  [
    "lg:flex",
    [
      [
        174,
        181,
      ],
    ],
  ],
  [
    "lg:mb-0",
    [
      [
        2034,
        2041,
      ],
    ],
  ],
  [
    "w-full",
    [
      [
        110,
        116,
      ],
      [
        232,
        238,
      ],
      [
        663,
        669,
      ],
    ],
  ],
  [
    "left-0",
    [
      [
        214,
        220,
      ],
      [
        646,
        652,
      ],
    ],
  ],
  [
    "lg:p-4",
    [
      [
        460,
        466,
      ],
    ],
  ],
  [
    "lg:p-0",
    [
      [
        902,
        908,
      ],
    ],
  ],
  [
    "top-0",
    [
      [
        221,
        226,
      ],
    ],
  ],
  [
    "gap-2",
    [
      [
        869,
        874,
      ],
    ],
  ],
  [
    "mb-32",
    [
      [
        2011,
        2016,
      ],
    ],
  ],
  [
    "p-24",
    [
      [
        82,
        86,
      ],
    ],
  ],
  [
    "z-10",
    [
      [
        105,
        109,
      ],
    ],
  ],
  [
    "pb-6",
    [
      [
        310,
        314,
      ],
    ],
  ],
  [
    "pt-8",
    [
      [
        315,
        319,
      ],
    ],
  ],
  [
    "h-48",
    [
      [
        658,
        662,
      ],
    ],
  ],
  [
    "mb-3",
    [
      [
        2011,
        2015,
      ],
      [
        2455,
        2459,
      ],
      [
        3165,
        3169,
      ],
      [
        3904,
        3908,
      ],
      [
        4622,
        4626,
      ],
    ],
  ],
  [
    "px-5",
    [
      [
        2257,
        2261,
      ],
      [
        2967,
        2971,
      ],
      [
        3706,
        3710,
      ],
      [
        4424,
        4428,
      ],
    ],
  ],
  [
    "py-4",
    [
      [
        2262,
        2266,
      ],
      [
        2972,
        2976,
      ],
      [
        3711,
        3715,
      ],
      [
        4429,
        4433,
      ],
    ],
  ],
  [
    "m-0",
    [
      [
        642,
        645,
      ],
      [
        2660,
        2663,
      ],
      [
        3371,
        3374,
      ],
      [
        4114,
        4117,
      ],
      [
        4829,
        4832,
      ],
    ],
  ],
  [
    "p-8",
    [
      [
        875,
        878,
      ],
    ],
  ],
]
`;
